File: Project/OutlinedTXTReportExporter.cpp
#include "OutlinedTXTReportExporter.hpp"

#include <sstream>

OutlinedTXTReportExporter::OutlinedTXTReportExporter(const std::string& filePath, const std::string& title, const std::string& body, bool includeDate) :
    TXTReportExporter(filePath, title, body, includeDate) {}

OutlinedTXTReportExporter::OutlinedTXTReportExporter(const OutlinedTXTReportExporter& other) :
    TXTReportExporter(other) {}

OutlinedTXTReportExporter& OutlinedTXTReportExporter::operator=(const TXTReportExporter& other) {
    m_filePath = other.GetFilePath();
    m_title = other.GetTitle();
    m_body = other.GetBody();
    m_includeDate = other.GetIncludeDate();
    return *this;
}


std::string OutlinedTXTReportExporter::RenderTable(const GroupTableData& data) {
    std::string table = TXTReportExporter::RenderTable(data);
    int width = table.find('\n');
    std::string outline = TABLE_CROSS_SYMBOL + std::string(width, TABLE_HORIZONTAL_SYMBOL) + TABLE_CROSS_SYMBOL;

    std::stringstream resultss;

    std::stringstream tabless(table);
    std::string buf;
    resultss << outline << std::endl;
    while (std::getline(tabless, buf)) {
        resultss << TABLE_VERTICAL_SYMBOL << buf << TABLE_VERTICAL_SYMBOL << std::endl;
    }
    resultss << outline;

    return resultss.str();
}

File: ProjectCS/OutlinedTXTReportExporter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SessionResultsCS
{
    public class OutlinedTXTReportExporter : TXTReportExporter
    {
        public OutlinedTXTReportExporter(string filePath, string? title = null, string? body = null, bool includeDate = false)
            : base(filePath, title, body, includeDate)
        {
        }

        public override string RenderTable(GroupTableData data)
        {
            string table = base.RenderTable(data);
            int width = table.IndexOf(Environment.NewLine);
            string outline = TABLE_CROSS_SYMBOL + new string(TABLE_HORIZONTAL_SYMBOL, width) + TABLE_CROSS_SYMBOL;

            StringBuilder builder = new StringBuilder();

            builder.AppendLine(outline);
            using (StringReader reader = new StringReader(table))
            {
                string? line;
                while ((line = reader.ReadLine()) != null) {
                    builder.Append(TABLE_VERTICAL_SYMBOL);
                    builder.Append(line);
                    builder.Append(TABLE_VERTICAL_SYMBOL);
                    builder.AppendLine();
                }
            }
            builder.Append(outline);

            return builder.ToString();
        }
    }
}
